src/utilities/oas-validator.ts

```diff
@@ -114,31 +114,103 @@ class OASValidator {
     path: string[],
   ): ValidationFailure[] {
     let failures: ValidationFailure[] = [];
+    const actualType = typeof actual;
+
+    const invalidSchemaErrors = this.checkInvalidSchemaErrors(
+      actual,
+      expected,
+      path,
+    );
+    if (invalidSchemaErrors) {
+      failures = [...failures, ...invalidSchemaErrors];
+      return failures;
+    }
+
+    const expectedTypeError = this.checkExpectedTypeError(
+      actual,
+      expected,
+      path,
+    );
+
+    if (expectedTypeError) {
+      failures = [...failures, expectedTypeError];
+      return failures;
+    }
+
+    const enumValueErrors = this.checkEnumValueError(actual, expected, path);
+    if (enumValueErrors.length > 0) {
+      failures = [...failures, ...enumValueErrors];
+    }
+
+    if (Array.isArray(actual)) {
+      const itemSchemaErrors = this.checkItemSchemaErrors(
+        actual,
+        expected,
+        path,
+      );
+      if (itemSchemaErrors) {
+        failures = [...failures, ...itemSchemaErrors];
+      }
+    } else if (actualType === 'object') {
+      const propertiesErrors = this.checkObjectPropertiesErrors(
+        actual,
+        expected,
+        path,
+      );
+      if (propertiesErrors) {
+        failures = [...failures, ...propertiesErrors];
+      }
+    }
+
+    return failures;
```

Comment:
i think we need to add the `path.pop()` before this return. The fact that no tests are failing makes me think we also need to add in tests for that, but maybe that's part of the tests you noticed where missing

Reply:
ended changing to a more functional style